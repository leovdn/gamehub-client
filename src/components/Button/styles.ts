import styled, { DefaultTheme, css } from 'styled-components'
import { ButtonProps } from '.'

type WrapperProps = { $hasIcon: boolean } & Pick<
  ButtonProps,
  'size' | '$fullwidth'
>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    height: 3rem;
    font-size: ${theme.font.sizes.xsmall};
  `,
  medium: (theme: DefaultTheme) => css`
    height: 4rem;
    font-size: ${theme.font.sizes.small};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.medium};
  `,
  large: (theme: DefaultTheme) => css`
    height: 5rem;
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} ${theme.spacings.xlarge};
  `,
  $fullwidth: () => css`
    width: 100%;
  `,
  withIcon: (theme: DefaultTheme) => css`
    svg {
      width: 1.5rem;

      & + span {
        margin-left: ${theme.spacings.xxsmall};
      }
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, size, $fullwidth, $hasIcon }) => css`
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: ${theme.spacings.xxsmall};
    border-radius: ${theme.border.radius};

    border: none;
    background: linear-gradient(180deg, #ff4f5f 0%, #f062c0 50%);
    color: ${theme.colors.white};
    text-decoration: none;
    cursor: pointer;

    &:hover {
      background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
    }

    ${!!size && wrapperModifiers[size](theme)}
    ${$fullwidth && wrapperModifiers.$fullwidth()}
    ${$hasIcon && wrapperModifiers.withIcon(theme)}
  `}
`
