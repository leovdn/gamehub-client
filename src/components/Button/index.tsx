import { AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  $fullwidth?: boolean
  icon?: JSX.Element
  minimal?: boolean
  as?: React.ElementType
} & ButtonTypes

const Button = ({
  children,
  size = 'medium',
  $fullwidth = false,
  icon,
  minimal = false,
  ...props
}: ButtonProps) => {
  return (
    <S.Wrapper
      size={size}
      $fullwidth={$fullwidth}
      $hasIcon={!!icon}
      minimal={minimal}
      {...props}
    >
      {icon}
      {!!children && <span>{children}</span>}
    </S.Wrapper>
  )
}

export default Button
