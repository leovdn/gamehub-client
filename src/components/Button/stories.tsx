import Button from '.'
import { Meta, StoryFn } from '@storybook/react'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    }
  }
} as Meta

export const Default: StoryFn = (args) => <Button {...args} />

Default.args = {
  children: 'Buy Now'
}
