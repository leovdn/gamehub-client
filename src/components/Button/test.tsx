import { screen } from '@testing-library/react'

import Button from './index'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from 'styled-icons/material-outlined'

describe('<Button />', () => {
  it('should render the medium button size by Default', () => {
    const { container } = renderWithTheme(<Button>Buy Now</Button>)

    expect(screen.getByRole('button', { name: /Buy Now/i })).toHaveStyle({
      height: '4rem',
      'font-size': '1.4rem',
      padding: '0.8rem 3.2rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small button size', () => {
    renderWithTheme(<Button size="small">Buy Now</Button>)

    expect(screen.getByRole('button', { name: /Buy Now/i })).toHaveStyle({
      height: '3rem',
      'font-size': '1.2rem'
    })
  })

  it('should render the large button size', () => {
    renderWithTheme(<Button size="large">Buy Now</Button>)

    expect(screen.getByRole('button', { name: /Buy Now/i })).toHaveStyle({
      height: '5rem',
      'font-size': '1.6rem',
      padding: '0.8rem 4.8rem'
    })
  })

  it('should render a full-width version', () => {
    renderWithTheme(
      <Button size="small" $fullwidth>
        Buy Now
      </Button>
    )

    expect(screen.getByRole('button', { name: /Buy Now/i })).toHaveStyle({
      width: '100%'
    })
  })

  it('should render an icon version', () => {
    renderWithTheme(
      <Button size="small" icon={<AddShoppingCart data-testid="icon" />}>
        Buy Now
      </Button>
    )

    expect(screen.getByText(/Buy Now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render the Button as a Link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy Now
      </Button>
    )

    expect(screen.getByRole('link', { name: /Buy Now/i })).toHaveAttribute('href', '/link')
  })
})
