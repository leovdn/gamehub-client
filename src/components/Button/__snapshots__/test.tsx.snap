// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium button size by Default 1`] = `
.c0 {
  padding: 0.8rem;
  border-radius: 0.4rem;
  border: none;
  background: linear-gradient(180deg,#ff4f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  cursor: pointer;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  background: linear-gradient(180deg,#e35565 0%,#d958a6 50%);
}

<button
  class="c0"
>
  <span>
    Buy Now
  </span>
</button>
`;
